pkg/lib/operatorstatus/handler.go

```diff
@@ -0,0 +1,118 @@
+package operatorstatus
+
+import (
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/csv"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+)
+
+const (
+	// SelectorKey is the key of the label we use to identify the
+	// corresponding ClusterServiceVersion object related to the cluster operator.
+	// If we want to update a cluster operator named "package-server" then the
+	// corresponding ClusterServiceVersion must have the following label
+	//
+	// "olm.clusteroperator.name": "package-server"
+	//
+	SelectorKey = "olm.clusteroperator.name"
+)
+
+// NewCSVWatchNotificationHandler returns a new instance of csv.WatchNotification
+// This can be used to get notification of every CSV reconciliation request.
+func NewCSVWatchNotificationHandler(log *logrus.Logger, csvSet csv.SetGenerator, finder csv.ReplaceFinder, sender Sender) *handler {
+	logger := log.WithField("monitor", "clusteroperator")
+	return &handler{
+		csvSet: csvSet,
+		finder: finder,
+		sender: sender,
+		logger: logger,
+	}
+}
+
+type handler struct {
+	csvSet csv.SetGenerator
+	finder csv.ReplaceFinder
+	sender Sender
+	logger *logrus.Entry
+}
+
+// OnAddOrUpdate is invoked when a CSV has been added or edited. We tap into
+// this notification and do the following:
+//
+// a. Make sure this is the CSV related to the cluster operator resource we are
+//    tracking. Otherwise, do nothing.
+// b. If this is the right CSV then send it to the monitor.
+func (h *handler) OnAddOrUpdate(in *v1alpha1.ClusterServiceVersion) {
+	name, matched := h.isMatchingCSV(in)
+	if !matched {
+		return
+	}
+
+	h.logger.Debugf("OnAddOrUpdate - found a matching CSV name=%s phase=%s, sending notification", in.GetName(), in.Status.Phase)
+
+	final := h.getLatestInReplacementChain(in)
```

Comment:
is there utility in tracking whether we're "final" or not?

can we just say "if there's a replacement, progressing=true, else progressing=false"?

Reply:
If a `csv` is being replaced then I set message for `Progressing` to
`Working toward {version of the final csv}"`

Thoughts?